pkg/cmd/pulumi/plugin_run.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+
+	"github.com/blang/semver"
+	"github.com/spf13/cobra"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+func newPluginRunCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:    "run KIND NAME[@VERSION] [ARGS]",
+		Args:   cmdutil.MinimumNArgs(2),
+		Hidden: !env.Dev.Value(),
+		Short:  "Run a command on a plugin binary",
+		Long: "Run a command on a plugin binary.\n" +
+			"\n" +
+			"Directly executes a plugin binary, if VERSION is not specified " +
+			"the latest installed plugin will be used.",
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			if !workspace.IsPluginKind(args[0]) {
+				return fmt.Errorf("unrecognized plugin kind: %s", args[0])
+			}
+			kind := workspace.PluginKind(args[0])
+
+			// Parse the name and version from the second argument in the form of "NAME[@VERSION]".
+			name, version, found := strings.Cut(args[1], "@")
+			if !tokens.IsName(name) {
+				return fmt.Errorf("invalid plugin name %q", name)
+			}
+			var sv *semver.Version
+			if found {
+				v, err := semver.ParseTolerant(version)
+				if err != nil {
+					return fmt.Errorf("invalid plugin version %q: %w", version, err)
+				}
+				sv = &v
+			}
```

Comment:
Optional: The number of variables that a reader has to be aware of here could be reduced by having this block produce only two things: name and a semver.Version.

Something like the following:

```suggestion
			name := args[1]
			var sv *semver.Version
			if namePart, vstr, ok := strings.Cut(name, "@"); ok {
				v, err := semver.ParseTolerant(vstr)
				if err != nil {
					return fmt.Errorf("invalid plugin version %q: %w", vstr, err)
				}
				name, sv = namePart, &v
			}
			if !tokens.IsName(name) {
				return fmt.Errorf("invalid plugin name %q", name)
			}
```

At that point, we're free to use "version" as the name for "vs" instead (which I didn't do to keep the suggestion apply-able). It also reduces the risk of accidentally referencing the wrong version variable.